app/lib/cms/s3.rb

```diff
@@ -26,38 +26,20 @@ def fips_environment?
       false
     end
 
-    def bucket
-      config.fetch(:bucket) if enabled?
-    end
-
-    def region
-      config.fetch(:region) if enabled?
-    end
-
     def credentials
       return unless enabled?
 
-      Aws::CredentialsService.call(config)
-    end
-
-    def hostname
-      config[:hostname].presence if enabled?
+      CMS::AwsCredentials.instance.credentials
```

Comment:
I don't quite like the approach of calling the `.credentials` method with no arguments, but then this method calls back some S3 methods, so it's a two-way dependency.
I think passing `config` with all of the available props makes sense and is easier - maybe pass the hash which is the result of `ALL_KEYS.map { |key| [key, S3.public_send(key)] }.to_h`, so we also don't need to call these public methods two times on initialization (I know, as it will only be executed once, it's not a big deal, but still...)

If `CMS::AwsCredentials.instance.credentials` was called from somewhere else, then it might make sense to skip the arguments, but as it's only being called from S3 module itself, I think we can pass the config.

Reply:
If we pass arguments, this means that it would need to be an abstraction for handling any credentials. But then if it process any credentials, It makes no sense to be singleton and cache. Because if arguments change, also cache would be invalid.

This is what I didn't like in previous implementation - that we pass arguments but then we cache anyway.

Alternatively I can make it a utility class and let caching happen within `CMS::S3` module itself. WDYT?

Reply:
I'd keep the Singleton approach.
`CMS::AwsCredentials.instance.credentials` fetches the values it needs from ENV variables (going through `CMS::S3` now) and [1] returns some values from those ENV variables if we're authenticating with IAM or [2] calls STS to get the STS credentials and caches the Client.
Being a singleton, we can be sure that the credentials will be the same all across the application and we can see it as part of the Application's configuration.

Reply:
OK, yes, makes sense :+1: